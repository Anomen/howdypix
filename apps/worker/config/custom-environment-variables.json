{
  "app": {
    "processing": {
      "nbThreads": "THREADS"
    },
    "rabbitMQ": {
      "url": "RABBITMQ_URL"
    }
  }
}
