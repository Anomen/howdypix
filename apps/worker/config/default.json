{
  "rabbitMq": {
    "url": "amqp://localhost"
  },
  "distThumbnails": "./thumbnails"
}
