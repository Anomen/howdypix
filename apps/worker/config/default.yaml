#####################################################
# Application settings
#####################################################
app:
  rabbitMQ:
    url: amqp://localhost