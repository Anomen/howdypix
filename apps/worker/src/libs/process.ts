import sharp from "sharp";
import mkdirp from "mkdirp";
import { join, parse } from "path";
import { statSync } from "fs";
import { ExifImage } from "exif";
import { ExifData, HFile, ProcessData, StatData } from "@howdypix/shared-types";
import {
  appDebug,
  generateThumbnailPaths,
  thumbnailPath,
  hfile2path
} from "@howdypix/utils";

export async function fetchExif(root: string, path: string): Promise<ExifData> {
  return new Promise((resolve, reject) => {
    // eslint-disable-next-line no-new
    new ExifImage(join(root, path), (error, data) => {
      if (!error) {
        resolve({
          ISO: data.exif.ISO,
          aperture: data.exif.ApertureValue,
          shutter: data.exif.ExposureTime,
          createDate: data.exif.CreateDate
            ? new Date(data.exif.CreateDate).getMilliseconds()
            : undefined,
          make: data.image.Make,
          model: data.image.Model
        });
      } else {
        reject(error);
      }
    });
  });
}

export async function fetchStat(root: string, path: string): Promise<StatData> {
  const absolutePath = join(root, path);
  const stat = statSync(absolutePath);
  return {
    inode: stat.ino,
    ctime: stat.ctimeMs,
    mtime: stat.mtimeMs,
    birthtime: stat.birthtimeMs,
    size: stat.size
  };
}

export async function createThumbnails(
  thumbnailsDir: string,
  root: string,
  hfile: HFile
): Promise<string[]> {
  const { dir } = parse(thumbnailPath(thumbnailsDir, hfile));

  mkdirp.sync(dir);

  return Promise.all(
    generateThumbnailPaths(thumbnailsDir, hfile).map(async data => {
      await sharp(join(root, hfile2path(hfile).toString()))
        .resize(data.width)
        .toFile(data.path);
      return data.path;
    })
  );
}

export async function process(
  thumbnailsDir: string,
  root: string,
  hfile: HFile
): Promise<ProcessData> {
  const path = hfile2path(hfile).toString();
  let stat: StatData;
  let exif: ExifData;

  try {
    stat = await fetchStat(root, path);
  } catch (e) {
    appDebug(e);
    stat = {
      birthtime: 0,
      ctime: 0,
      inode: 0,
      mtime: 0,
      size: 0
    };
  }

  try {
    exif = await fetchExif(root, path);
  } catch (e) {
    appDebug(e);
    exif = {};
  }

  const thumbnails = await createThumbnails(thumbnailsDir, root, hfile);

  const data = {
    exif,
    stat,
    thumbnails,
    hfile,
    root
  };

  appDebug("processed")(`${path}: ${JSON.stringify(data)}`);

  return data;
}
