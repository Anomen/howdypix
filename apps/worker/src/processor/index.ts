import { FileProcessor } from "./types";
import imageProcessor from "./image";

export function getFileProcessor(mime: string | null): FileProcessor | null {
  switch (mime) {
    case "image/jpeg":
    case "image/png":
      return imageProcessor;

    default:
      return null;
  }
}
