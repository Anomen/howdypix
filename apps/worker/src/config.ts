import config from "config";

export default {
  rabbitMq: {
    url: (config.get("app.rabbitMQ.url") as string) || "amqp://localhost"
  }
};
