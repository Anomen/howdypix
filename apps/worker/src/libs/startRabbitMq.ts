import { connect as connectRabbitMq } from "amqplib";
import { process } from "./process";
import { MessageProcess } from "@howdypix/shared-types";

const queueName = "toProcess";

export async function startRabbitMq(url: string) {
  const connection = await connectRabbitMq(url);
  const channel = await connection.createChannel();

  await channel.assertQueue(queueName);

  await channel.consume(queueName, async msg => {
    if (msg) {
      const data: MessageProcess = JSON.parse(msg.content.toString());
      console.log(await process(data.root, data.path));
      channel.ack(msg);
    }
  });

  return channel;
}
