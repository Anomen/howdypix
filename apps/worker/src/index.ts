import config from "./config";
import { startRabbitMq } from "./libs/startRabbitMq";

async function main() {
  await startRabbitMq(config.rabbitMq.url);
}

main();
