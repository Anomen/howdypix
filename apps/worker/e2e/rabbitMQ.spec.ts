import { consume, sendToQueue } from "@howdypix/utils";
import { connectToRabbitMq } from "@howdypix/utils/dist/rabbitMq";
import { HFile, QueueName } from "@howdypix/shared-types";
import { Channel, Connection } from "amqplib";
import waitForExpect from "wait-for-expect";
import { startRabbitMq } from "../src/libs/startRabbitMq";
import { process } from "../src/libs/process";
import { appConfig } from "../src/config";

jest.mock("../src/libs/process", () => ({
  process: jest.fn(),
}));

describe("startRabbitMq", () => {
  let connection: Connection;
  let channel: Channel;

  const baseHfile: HFile = {
    dir: "dir",
    file: "file",
    source: "source",
    name: "name",
  };

  beforeAll(async () => {
    connection = await connectToRabbitMq(appConfig.rabbitMQ.url);
    channel = await connection.createChannel();

    // Remove all queues
    try {
      await channel.deleteQueue(QueueName.TO_PROCESS);
      await channel.deleteQueue(QueueName.PROCESSED);
    } catch (e) {
      // do nothing
    }

    // Create new queues
    try {
      await channel.assertQueue(QueueName.TO_PROCESS);
      await channel.assertQueue(QueueName.PROCESSED);
    } catch (e) {
      // do nothing
    }
  });

  beforeEach(async () => {
    try {
      await channel.purgeQueue(QueueName.PROCESSED);
      await channel.purgeQueue(QueueName.TO_PROCESS);
    } catch (e) {
      // do nothing
    }
  });

  afterAll(async () => {
    await channel.close();
    await connection.close();
  });

  describe("when receiving a message from RabbitMQ", () => {
    beforeEach(async () => {
      // Initialize
      (process as jest.Mock).mockImplementation(() => {
        return Promise.resolve({ data: 1 });
      });

      // Run
      await startRabbitMq(appConfig.rabbitMQ);
      await sendToQueue(channel, QueueName.TO_PROCESS, {
        thumbnailsDir: "thumbnailsDir",
        hfile: baseHfile,
        root: "/root",
      });
    });

    it("should process the file when receiving a message from RabbitMQ", async () => {
      // Test
      await waitForExpect(() => {
        expect(process).toHaveBeenCalled();
      });
      expect((process as jest.Mock).mock.calls[0]).toMatchSnapshot();
    });

    it("should send a message back to RabbitMQ", async (done) => {
      consume(channel, QueueName.PROCESSED, (msg) => {
        waitForExpect(() => {
          expect(msg?.data).toMatchSnapshot();
        }).then(done);
      });
    });
  });

  it("when having an error processing, should not fail", async () => {
    // Initialize
    (process as jest.Mock).mockImplementation(() => {
      return Promise.reject(new Error("error"));
    });

    jest.spyOn(console, "error").mockImplementationOnce(() => {});

    // Run
    await startRabbitMq(appConfig.rabbitMQ);
    await sendToQueue(channel, QueueName.TO_PROCESS, {
      thumbnailsDir: "thumbnailsDir",
      hfile: baseHfile,
      root: "/root",
    });

    // Test
    await waitForExpect(() => {
      expect(process).toHaveBeenCalled();
      expect(console.error).toHaveBeenCalled();
    });
  });
});
