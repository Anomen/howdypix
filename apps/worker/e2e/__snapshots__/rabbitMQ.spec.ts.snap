// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startRabbitMq when receiving a message from RabbitMQ should process the file when receiving a message from RabbitMQ 1`] = `
Array [
  "thumbnailsDir",
  "/root",
  Object {
    "dir": "dir",
    "file": "file",
    "name": "name",
    "source": "source",
  },
]
`;

exports[`startRabbitMq when receiving a message from RabbitMQ should send a message back to RabbitMQ 1`] = `
Object {
  "data": 1,
}
`;
